lib/ansible/modules/packaging/os/dnf.py

```diff
@@ -303,6 +304,30 @@ def __init__(self, module):
         self._ensure_dnf()
         self.lockfile = "/var/cache/dnf/*_lock.pid"
         self.pkg_mgr_name = "dnf"
+        self.base = self._base()
+        self.exclude_pkgnames = self._get_exclude_pkgnames()
+
+    def _pattern_to_pkgname(self, base, pattern):
+        if dnf.util.is_glob_pattern(pattern):
+            q = base.sack.query().filter(name__glob=pattern)
+            return map(lambda p: p.name, q)
+        else:
+            return (pattern.name,)
+
+    def _get_exclude_pkgnames(self):
+        orig_disable_excludes = self.base.conf.disable_excludes
+        self.base.conf.disable_excludes = ['all']
+        self.base.fill_sack()
+
+        expanded_excludes = [self._pattern_to_pkgname(self.base, p) for p in self.base.conf.exclude]
+
+        if orig_disable_excludes:
+            self.base.conf.disable_excludes = orig_disable_excludes
+        else:
+            self.base.conf.disable_excludes = []
+        self.base.fill_sack()
+
+        return itertools.chain.from_iterable(expanded_excludes)
```

Comment:
Going through itertools here ends up creating both a list and a generator so it's likely slower than just returning a list.  If the list is very large and your goal is to reduce memory consumption, you can use a generator expresssion instead of a list comprehension on line 322:
``` python
expanded_excludes = (self._pattern_to_pkgname(self.base, p) for p in self.base.conf.exclude)
```

However, for iteration over most data, lists are faster than other iterables so it is a time-space tradeoff.

Reply:
Good to know, also this ends up needing to be a list anyways because once the `dnf.query.Query` object goes out of scope, the generator returns nothing. Thanks!